packages/onft-evm/contracts/onft721/ONFT721Core.sol

```diff
@@ -0,0 +1,166 @@
+        SendParam calldata _sendParam,
+        MessagingFee calldata _fee,
+        address _refundAddress
+    ) external payable virtual returns (MessagingReceipt memory msgReceipt) {
+        if (_sendParam.to == bytes32(0)) revert InvalidReceiver();
+        _debit(msg.sender, _sendParam.tokenId, _sendParam.dstEid);
+
+        (bytes memory message, bytes memory options) = _buildMsgAndOptions(_sendParam);
+
+        // @dev Sends the message to the LayerZero Endpoint, returning the MessagingReceipt.
+        msgReceipt = _lzSend(_sendParam.dstEid, message, options, _fee, _refundAddress);
+        emit ONFTSent(msgReceipt.guid, _sendParam.dstEid, msg.sender, _sendParam.tokenId);
+    }
+
+    /// @dev Internal function to build the message and options.
+    /// @param _sendParam The parameters for the send() operation.
+    /// @return message The encoded message.
+    /// @return options The encoded options.
+    function _buildMsgAndOptions(
+        SendParam calldata _sendParam
+    ) internal view virtual returns (bytes memory message, bytes memory options) {
+        bool hasCompose;
+        (message, hasCompose) = ONFT721MsgCodec.encode(_sendParam.to, _sendParam.tokenId, _sendParam.composeMsg);
+        uint16 msgType = hasCompose ? SEND_AND_COMPOSE : SEND;
+
+        options = combineOptions(_sendParam.dstEid, msgType, _sendParam.extraOptions);
+
+        // @dev Optionally inspect the message and options depending if the OApp owner has set a msg inspector.
+        // @dev If it fails inspection, needs to revert in the implementation. ie. does not rely on return boolean
+        if (msgInspector != address(0)) IOAppMsgInspector(msgInspector).inspect(message, options);
+    }
+
+    /// @dev Internal function to handle the receive on the LayerZero endpoint.
+    /// @param _origin The origin information.
+    ///  - srcEid: The source chain endpoint ID.
+    ///  - sender: The sender address from the src chain.
+    ///  - nonce: The nonce of the LayerZero message.
+    /// @param _guid The unique identifier for the received LayerZero message.
+    /// @param _message The encoded message.
+    /// @dev _executor The address of the executor.
+    /// @dev _extraData Additional data.
+    function _lzReceive(
+        Origin calldata _origin,
+        bytes32 _guid,
+        bytes calldata _message,
+        address /*_executor*/, // @dev unused in the default implementation.
+        bytes calldata /*_extraData*/ // @dev unused in the default implementation.
+    ) internal virtual override {
+        address toAddress = _message.sendTo().bytes32ToAddress();
+        uint256 tokenId = _message.tokenId();
+
+        _credit(toAddress, tokenId, _origin.srcEid);
+
+        if (_message.isComposed()) {
+            bytes memory composeMsg = ONFTComposeMsgCodec.encode(_origin.nonce, _origin.srcEid, _message.composeMsg());
+            endpoint.sendCompose(toAddress, _guid, 0 /* the index of composed message*/, composeMsg);
```

Comment:
Elaborate this comment, as there is only ever one "compose index" for each lzReceive, thus its hardcoded. If batching is implemented, will need to have this correspond to that index